File: packages/shared/src/skills.ts
import { KICK_RANGE_PX, MELEE_ATTACK_COOLDOWN_SECONDS, PLAYER_BASE_AIR_CONTROL, PLAYER_BASE_ATTACK_SPEED, PLAYER_BASE_DAMAGE, PLAYER_BASE_JUMP_POWER } from "./constants.js";
import type { PlayerId, PlayerState } from "./types.js";

export type SkillCategory = "attack" | "defense" | "mobility" | "utility";
export type SkillRarity = "common" | "rare" | "epic";

export interface SkillCardDefinition {
  id: string;
  name: string;
  category: SkillCategory;
  rarity: SkillRarity;
  description: string;
  icon: string;
  maxStacks: number;
  weight: number;
}

export interface SkillCard {
  id: string;
  name: string;
  category: SkillCategory;
  rarity: SkillRarity;
  description: string;
  icon: string;
  maxStacks: number;
  currentStacks: number;
}

export interface SkillCardOffer {
  offerId: string;
  playerId: PlayerId;
  cards: SkillCard[];
}

export const SKILL_BALANCE = {
  longReachPerStack: 0.15,
  maxHitRangeMultiplier: 1.75,
  heavyHitPerStack: 0.10,
  maxAtkMultiplier: 1.75,
  quickStrikePerStack: 0.08,
  minAttackCooldownMultiplier: 0.65,
  protectiveAuraReductionPerStack: 0.08,
  maxDamageReduction: 0.50,
  shieldBase: 25,
  shieldPerStack: 15,
  shieldRegenPerSecond: 5,
  shieldRegenDelayMs: 3000,
  springLegsPerStack: 0.10,
  maxJumpMultiplier: 1.40,
  airControlPerStack: 0.12,
  maxAirControlMultiplier: 1.50,
  dashBaseCooldownMs: 1800,
  dashCooldownReductionPerStack: 0.12,
  pickupRadiusPerStack: 0.25,
  maxPickupRadiusMultiplier: 2.5,
  xpCollectibleBonusPerStack: 0.20,
  killXpBonusPerStack: 0.15,
} as const;

export const SKILL_REGISTRY: readonly SkillCardDefinition[] = [
  { id: "long_reach", name: "Long Reach", category: "attack", rarity: "common", description: "Hit range +15%.", icon: "ARC", maxStacks: 5, weight: 16 },
  { id: "heavy_hit", name: "Heavy Hit", category: "attack", rarity: "common", description: "ATK damage +10%.", icon: "FIST", maxStacks: 5, weight: 16 },
  { id: "quick_strike", name: "Quick Strike", category: "attack", rarity: "rare", description: "Attack cooldown -8%.", icon: "SPD", maxStacks: 4, weight: 11 },
  { id: "shockwave_hit", name: "Shockwave Hit", category: "attack", rarity: "epic", description: "Every few hits emits a pixel shockwave.", icon: "WAVE", maxStacks: 4, weight: 6 },
  { id: "protective_aura", name: "Protective Aura", category: "defense", rarity: "common", description: "Incoming damage -8%.", icon: "AURA", maxStacks: 4, weight: 14 },
  { id: "energy_shield", name: "Energy Shield", category: "defense", rarity: "rare", description: "Adds a regenerating shield before HP.", icon: "SHLD", maxStacks: 5, weight: 11 },
  { id: "last_stand", name: "Last Stand", category: "defense", rarity: "rare", description: "Low HP grants extra damage reduction.", icon: "LAST", maxStacks: 4, weight: 9 },
  { id: "spring_legs", name: "Spring Legs", category: "mobility", rarity: "common", description: "Jump power +10%.", icon: "JUMP", maxStacks: 4, weight: 14 },
  { id: "air_control", name: "Air Control", category: "mobility", rarity: "common", description: "Air steering +12%.", icon: "WIND", maxStacks: 4, weight: 13 },
  { id: "double_jump", name: "Double Jump", category: "mobility", rarity: "epic", description: "Unlocks an extra air jump.", icon: "DBL", maxStacks: 3, weight: 6 },
  { id: "dash", name: "Dash", category: "mobility", rarity: "epic", description: "Unlocks a short horizontal dash.", icon: "DASH", maxStacks: 4, weight: 6 },
  { id: "xp_magnet", name: "XP Magnet", category: "utility", rarity: "common", description: "Pickup radius +25%.", icon: "MAG", maxStacks: 4, weight: 14 },
  { id: "dark_harvest", name: "Dark Harvest", category: "utility", rarity: "rare", description: "XP collectibles give +20% XP.", icon: "XP+", maxStacks: 4, weight: 10 },
  { id: "vital_growth", name: "Vital Growth", category: "utility", rarity: "common", description: "Max HP +15.", icon: "HP+", maxStacks: 6, weight: 12 },
  { id: "battle_learning", name: "Battle Learning", category: "utility", rarity: "rare", description: "Enemy kill XP +15%.", icon: "BOOK", maxStacks: 4, weight: 10 },
] as const;

export const SKILL_BY_ID = new Map(SKILL_REGISTRY.map((skill) => [skill.id, skill]));

function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, value));
}

function diminishingBonus(stacks: number, firstStackValue: number, decay: number): number {
  let total = 0;
  for (let i = 0; i < stacks; i += 1) total += firstStackValue / (1 + i * decay);
  return total;
}

function hashString(value: string): number {
  let hash = 2166136261;
  for (let i = 0; i < value.length; i += 1) {
    hash ^= value.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0;
}

function nextSeed(seed: number): number {
  return (Math.imul(seed, 1664525) + 1013904223) >>> 0;
}

function random01(seedRef: { value: number }): number {
  seedRef.value = nextSeed(seedRef.value);
  return seedRef.value / 0x1_0000_0000;
}

export function ensurePlayerSkillState(player: PlayerState): void {
  player.hitRange = Number.isFinite(player.hitRange) ? player.hitRange : KICK_RANGE_PX;
  player.attackCooldownMs = Number.isFinite(player.attackCooldownMs) ? player.attackCooldownMs : MELEE_ATTACK_COOLDOWN_SECONDS * 1000;
  player.damageReduction = Number.isFinite(player.damageReduction) ? player.damageReduction : 0;
  player.shield = Number.isFinite(player.shield) ? player.shield : 0;
  player.maxShield = Number.isFinite(player.maxShield) ? player.maxShield : 0;
  player.shieldRegenPerSecond = Number.isFinite(player.shieldRegenPerSecond) ? player.shieldRegenPerSecond : 0;
  player.shieldRegenDelayMs = Number.isFinite(player.shieldRegenDelayMs) ? player.shieldRegenDelayMs : SKILL_BALANCE.shieldRegenDelayMs;
  player.lastDamageAt = Number.isFinite(player.lastDamageAt) ? player.lastDamageAt : 0;
  player.shieldRegenCooldownMs = Number.isFinite(player.shieldRegenCooldownMs) ? player.shieldRegenCooldownMs : 0;
  player.jumpPowerMultiplier = Number.isFinite(player.jumpPowerMultiplier) ? player.jumpPowerMultiplier : 1;
  player.airControlMultiplier = Number.isFinite(player.airControlMultiplier) ? player.airControlMultiplier : 1;
  player.extraJumps = Number.isFinite(player.extraJumps) ? player.extraJumps : 0;
  player.extraJumpsUsed = Number.isFinite(player.extraJumpsUsed) ? player.extraJumpsUsed : 0;
  player.dashUnlocked = !!player.dashUnlocked;
  player.dashCooldownMs = Number.isFinite(player.dashCooldownMs) ? player.dashCooldownMs : SKILL_BALANCE.dashBaseCooldownMs;
  player.dashCooldownRemainingMs = Number.isFinite(player.dashCooldownRemainingMs) ? player.dashCooldownRemainingMs : 0;
  player.dashTimerMs = Number.isFinite(player.dashTimerMs) ? player.dashTimerMs : 0;
  player.pickupRadius = Number.isFinite(player.pickupRadius) ? player.pickupRadius : 24;
  player.xpGainMultiplier = Number.isFinite(player.xpGainMultiplier) ? player.xpGainMultiplier : 1;
  player.killXpMultiplier = Number.isFinite(player.killXpMultiplier) ? player.killXpMultiplier : 1;
  player.selectedSkills = player.selectedSkills && typeof player.selectedSkills === "object" ? player.selectedSkills : {};
  player.shockwaveCounter = Number.isFinite(player.shockwaveCounter) ? player.shockwaveCounter : 0;
}

export function skillStacks(player: PlayerState, skillId: string): number {
  ensurePlayerSkillState(player);
  return Math.max(0, Math.floor(player.selectedSkills[skillId] ?? 0));
}

export function recalculateSkillDerivedStats(player: PlayerState): void {
  ensurePlayerSkillState(player);
  const longReach = skillStacks(player, "long_reach");
  const heavyHit = skillStacks(player, "heavy_hit");
  const quickStrike = skillStacks(player, "quick_strike");
  const protectiveAura = skillStacks(player, "protective_aura");
  const energyShield = skillStacks(player, "energy_shield");
  const springLegs = skillStacks(player, "spring_legs");
  const airControl = skillStacks(player, "air_control");
  const doubleJump = skillStacks(player, "double_jump");
  const dash = skillStacks(player, "dash");
  const xpMagnet = skillStacks(player, "xp_magnet");
  const darkHarvest = skillStacks(player, "dark_harvest");
  const battleLearning = skillStacks(player, "battle_learning");

  const baseDamage = Math.min(1.45, PLAYER_BASE_DAMAGE + diminishingBonus(player.relicFragments, 0.035, 0.32));
  const baseAttackSpeed = Math.min(1.28, PLAYER_BASE_ATTACK_SPEED + diminishingBonus(player.relicFragments, 0.014, 0.36));
  const baseJumpPower = Math.min(1.22, PLAYER_BASE_JUMP_POWER + diminishingBonus(player.crystals, 0.018, 0.42));
  const baseAirControl = Math.min(1.1, PLAYER_BASE_AIR_CONTROL + diminishingBonus(player.crystals, 0.007, 0.5));

  player.hitRange = Math.round(KICK_RANGE_PX * Math.min(SKILL_BALANCE.maxHitRangeMultiplier, 1 + longReach * SKILL_BALANCE.longReachPerStack));
  const atkMultiplier = Math.min(SKILL_BALANCE.maxAtkMultiplier, 1 + heavyHit * SKILL_BALANCE.heavyHitPerStack);
  player.damage = baseDamage * atkMultiplier;
  const cooldownMultiplier = Math.max(SKILL_BALANCE.minAttackCooldownMultiplier, 1 - quickStrike * SKILL_BALANCE.quickStrikePerStack);
  player.attackCooldownMs = MELEE_ATTACK_COOLDOWN_SECONDS * 1000 * cooldownMultiplier;
  player.attackSpeed = Math.max(baseAttackSpeed, baseAttackSpeed / cooldownMultiplier);
  player.damageReduction = Math.min(SKILL_BALANCE.maxDamageReduction, protectiveAura * SKILL_BALANCE.protectiveAuraReductionPerStack);
  if (energyShield > 0) {
    const maxShield = SKILL_BALANCE.shieldBase + Math.max(0, energyShield - 1) * SKILL_BALANCE.shieldPerStack;
    player.maxShield = maxShield;
    player.shield = clamp(player.shield, 0, maxShield);
    if (player.shield <= 0 && player.lastDamageAt === 0) player.shield = maxShield;
    player.shieldRegenPerSecond = SKILL_BALANCE.shieldRegenPerSecond;
    player.shieldRegenDelayMs = SKILL_BALANCE.shieldRegenDelayMs;
  } else {
    player.maxShield = 0;
    player.shield = 0;
    player.shieldRegenPerSecond = 0;
  }
  player.jumpPowerMultiplier = Math.min(SKILL_BALANCE.maxJumpMultiplier, 1 + springLegs * SKILL_BALANCE.springLegsPerStack);
  player.airControlMultiplier = Math.min(SKILL_BALANCE.maxAirControlMultiplier, 1 + airControl * SKILL_BALANCE.airControlPerStack);
  player.jumpPower = baseJumpPower * player.jumpPowerMultiplier;
  player.airControl = baseAirControl * player.airControlMultiplier;
  player.extraJumps = doubleJump > 0 ? 1 : 0;
  player.dashUnlocked = dash > 0;
  player.dashCooldownMs = SKILL_BALANCE.dashBaseCooldownMs * Math.max(0.45, 1 - Math.max(0, dash - 1) * SKILL_BALANCE.dashCooldownReductionPerStack);
  player.pickupRadius = 24 * Math.min(SKILL_BALANCE.maxPickupRadiusMultiplier, 1 + xpMagnet * SKILL_BALANCE.pickupRadiusPerStack);
  player.xpGainMultiplier = 1 + darkHarvest * SKILL_BALANCE.xpCollectibleBonusPerStack;
  player.killXpMultiplier = 1 + battleLearning * SKILL_BALANCE.killXpBonusPerStack;
}

export function applySkillToPlayer(player: PlayerState, skillId: string): boolean {
  const definition = SKILL_BY_ID.get(skillId);
  if (!definition) return false;
  ensurePlayerSkillState(player);
  const current = skillStacks(player, skillId);
  if (current >= definition.maxStacks) return false;
  player.selectedSkills[skillId] = current + 1;
  if (skillId === "vital_growth") {
    player.maxHealth += 15;
    player.health = Math.min(player.maxHealth, player.health + 15);
  }
  recalculateSkillDerivedStats(player);
  return true;
}

function availableSkills(player: PlayerState): SkillCardDefinition[] {
  ensurePlayerSkillState(player);
  return SKILL_REGISTRY.filter((skill) => skillStacks(player, skill.id) < skill.maxStacks);
}

function toCard(player: PlayerState, definition: SkillCardDefinition): SkillCard {
  return {
    id: definition.id,
    name: definition.name,
    category: definition.category,
    rarity: definition.rarity,
    description: definition.description,
    icon: definition.icon,
    maxStacks: definition.maxStacks,
    currentStacks: skillStacks(player, definition.id),
  };
}

export function generateSkillCardOffer(player: PlayerState, offerId: string, seedInput: string | number): SkillCardOffer | null {
  ensurePlayerSkillState(player);
  const pool = availableSkills(player);
  if (pool.length === 0) return null;
  const seedRef = { value: typeof seedInput === "number" ? seedInput >>> 0 : hashString(seedInput) };
  const chosen: SkillCardDefinition[] = [];
  const categories = new Set<SkillCategory>();

  while (chosen.length < Math.min(3, pool.length)) {
    const remaining = pool.filter((skill) => !chosen.some((card) => card.id === skill.id));
    const varied = remaining.filter((skill) => !categories.has(skill.category));
    const source = varied.length > 0 ? varied : remaining;
    const totalWeight = source.reduce((sum, skill) => sum + skill.weight, 0);
    let pick = random01(seedRef) * totalWeight;
    let selected = source[source.length - 1]!;
    for (const skill of source) {
      pick -= skill.weight;
      if (pick <= 0) {
        selected = skill;
        break;
      }
    }
    chosen.push(selected);
    categories.add(selected.category);
  }

  return {
    offerId,
    playerId: player.id,
    cards: chosen.map((definition) => toCard(player, definition)),
  };
}

export function pendingSkillVisuals(player: PlayerState): {
  shield: boolean;
  aura: boolean;
  dash: boolean;
  doubleJump: boolean;
  xpAura: boolean;
} {
  return {
    shield: skillStacks(player, "energy_shield") > 0,
    aura: skillStacks(player, "protective_aura") > 0 || skillStacks(player, "last_stand") > 0,
    dash: skillStacks(player, "dash") > 0,
    doubleJump: skillStacks(player, "double_jump") > 0,
    xpAura: skillStacks(player, "xp_magnet") > 0 || skillStacks(player, "dark_harvest") > 0,
  };
}

File: apps/server-go/internal/game/skills.go
package game

import "math"

const (
	skillLongReachPerStack              = 0.15
	skillMaxHitRangeMultiplier          = 1.75
	skillHeavyHitPerStack               = 0.10
	skillMaxAtkMultiplier               = 1.75
	skillQuickStrikePerStack            = 0.08
	skillMinAttackCooldownMultiplier    = 0.65
	skillProtectiveAuraReductionPerStack = 0.08
	skillMaxDamageReduction             = 0.50
	skillShieldBase                     = 25.0
	skillShieldPerStack                 = 15.0
	skillShieldRegenPerSecond           = 5.0
	skillShieldRegenDelayMs             = 3000.0
	skillSpringLegsPerStack             = 0.10
	skillMaxJumpMultiplier              = 1.40
	skillAirControlPerStack             = 0.12
	skillMaxAirControlMultiplier        = 1.50
	skillDashBaseCooldownMs             = 1800.0
	skillDashCooldownReductionPerStack  = 0.12
	skillPickupRadiusPerStack           = 0.25
	skillMaxPickupRadiusMultiplier      = 2.5
	skillXPCollectibleBonusPerStack     = 0.20
	skillKillXPBonusPerStack            = 0.15
)

type skillDefinition struct {
	id          string
	name        string
	category    string
	rarity      string
	description string
	icon        string
	maxStacks   int
	weight      int
}

var skillRegistry = []skillDefinition{
	{"long_reach", "Long Reach", "attack", "common", "Hit range +15%.", "ARC", 5, 16},
	{"heavy_hit", "Heavy Hit", "attack", "common", "ATK damage +10%.", "FIST", 5, 16},
	{"quick_strike", "Quick Strike", "attack", "rare", "Attack cooldown -8%.", "SPD", 4, 11},
	{"shockwave_hit", "Shockwave Hit", "attack", "epic", "Every few hits emits a pixel shockwave.", "WAVE", 4, 6},
	{"protective_aura", "Protective Aura", "defense", "common", "Incoming damage -8%.", "AURA", 4, 14},
	{"energy_shield", "Energy Shield", "defense", "rare", "Adds a regenerating shield before HP.", "SHLD", 5, 11},
	{"last_stand", "Last Stand", "defense", "rare", "Low HP grants extra damage reduction.", "LAST", 4, 9},
	{"spring_legs", "Spring Legs", "mobility", "common", "Jump power +10%.", "JUMP", 4, 14},
	{"air_control", "Air Control", "mobility", "common", "Air steering +12%.", "WIND", 4, 13},
	{"double_jump", "Double Jump", "mobility", "epic", "Unlocks an extra air jump.", "DBL", 3, 6},
	{"dash", "Dash", "mobility", "epic", "Unlocks a short horizontal dash.", "DASH", 4, 6},
	{"xp_magnet", "XP Magnet", "utility", "common", "Pickup radius +25%.", "MAG", 4, 14},
	{"dark_harvest", "Dark Harvest", "utility", "rare", "XP collectibles give +20% XP.", "XP+", 4, 10},
	{"vital_growth", "Vital Growth", "utility", "common", "Max HP +15.", "HP+", 6, 12},
	{"battle_learning", "Battle Learning", "utility", "rare", "Enemy kill XP +15%.", "BOOK", 4, 10},
}

func skillByID(id string) (skillDefinition, bool) {
	for _, skill := range skillRegistry {
		if skill.id == id {
			return skill, true
		}
	}
	return skillDefinition{}, false
}

func ensurePlayerSkillState(player *PlayerState) {
	if player.HitRange <= 0 {
		player.HitRange = KickRangePX
	}
	if player.AttackCooldownMs <= 0 {
		player.AttackCooldownMs = KickCooldownSeconds * 1000
	}
	if player.ShieldRegenDelayMs <= 0 {
		player.ShieldRegenDelayMs = skillShieldRegenDelayMs
	}
	if player.JumpPowerMultiplier <= 0 {
		player.JumpPowerMultiplier = 1
	}
	if player.AirControlMultiplier <= 0 {
		player.AirControlMultiplier = 1
	}
	if player.DashCooldownMs <= 0 {
		player.DashCooldownMs = skillDashBaseCooldownMs
	}
	if player.PickupRadius <= 0 {
		player.PickupRadius = PickupRadius
	}
	if player.XPGainMultiplier <= 0 {
		player.XPGainMultiplier = 1
	}
	if player.KillXPMultiplier <= 0 {
		player.KillXPMultiplier = 1
	}
	if player.SelectedSkills == nil {
		player.SelectedSkills = map[string]int{}
	}
}

func skillStacks(player PlayerState, skillID string) int {
	if player.SelectedSkills == nil {
		return 0
	}
	return max(0, player.SelectedSkills[skillID])
}

func recalculateSkillDerivedStats(player *PlayerState) {
	ensurePlayerSkillState(player)
	longReach := skillStacks(*player, "long_reach")
	heavyHit := skillStacks(*player, "heavy_hit")
	quickStrike := skillStacks(*player, "quick_strike")
	protectiveAura := skillStacks(*player, "protective_aura")
	energyShield := skillStacks(*player, "energy_shield")
	springLegs := skillStacks(*player, "spring_legs")
	airControl := skillStacks(*player, "air_control")
	doubleJump := skillStacks(*player, "double_jump")
	dash := skillStacks(*player, "dash")
	xpMagnet := skillStacks(*player, "xp_magnet")
	darkHarvest := skillStacks(*player, "dark_harvest")
	battleLearning := skillStacks(*player, "battle_learning")

	baseDamage := float64(1 + player.RelicFragments/8)
	atkMultiplier := math.Min(skillMaxAtkMultiplier, 1+float64(heavyHit)*skillHeavyHitPerStack)
	player.Damage = max(1, int(math.Round(baseDamage*atkMultiplier)))
	baseAttackSpeed := math.Min(3, 1+float64(player.RelicFragments)*0.014)
	cooldownMultiplier := math.Max(skillMinAttackCooldownMultiplier, 1-float64(quickStrike)*skillQuickStrikePerStack)
	player.AttackCooldownMs = KickCooldownSeconds * 1000 * cooldownMultiplier
	player.AttackSpeed = math.Max(baseAttackSpeed, baseAttackSpeed/cooldownMultiplier)
	player.HitRange = KickRangePX * math.Min(skillMaxHitRangeMultiplier, 1+float64(longReach)*skillLongReachPerStack)
	player.DamageReduction = math.Min(skillMaxDamageReduction, float64(protectiveAura)*skillProtectiveAuraReductionPerStack)
	if energyShield > 0 {
		maxShield := skillShieldBase + float64(max(0, energyShield-1))*skillShieldPerStack
		player.MaxShield = maxShield
		if player.Shield <= 0 && player.LastDamageAt == 0 {
			player.Shield = maxShield
		}
		player.Shield = math.Max(0, math.Min(player.Shield, maxShield))
		player.ShieldRegenPerSec = skillShieldRegenPerSecond
		player.ShieldRegenDelayMs = skillShieldRegenDelayMs
	} else {
		player.Shield = 0
		player.MaxShield = 0
		player.ShieldRegenPerSec = 0
	}
	player.JumpPowerMultiplier = math.Min(skillMaxJumpMultiplier, 1+float64(springLegs)*skillSpringLegsPerStack)
	player.AirControlMultiplier = math.Min(skillMaxAirControlMultiplier, 1+float64(airControl)*skillAirControlPerStack)
	player.JumpPower = math.Min(1.8, 1+float64(player.Crystals/3)*0.06) * player.JumpPowerMultiplier
	player.AirControl = math.Min(2, 1+float64(player.Crystals/3)*0.05) * player.AirControlMultiplier
	if doubleJump > 0 {
		player.ExtraJumps = 1
	} else {
		player.ExtraJumps = 0
	}
	player.DashUnlocked = dash > 0
	player.DashCooldownMs = skillDashBaseCooldownMs * math.Max(0.45, 1-float64(max(0, dash-1))*skillDashCooldownReductionPerStack)
	player.PickupRadius = PickupRadius * math.Min(skillMaxPickupRadiusMultiplier, 1+float64(xpMagnet)*skillPickupRadiusPerStack)
	player.XPGainMultiplier = 1 + float64(darkHarvest)*skillXPCollectibleBonusPerStack
	player.KillXPMultiplier = 1 + float64(battleLearning)*skillKillXPBonusPerStack
}

func applySkillToPlayer(player *PlayerState, skillID string) bool {
	def, ok := skillByID(skillID)
	if !ok {
		return false
	}
	ensurePlayerSkillState(player)
	current := skillStacks(*player, skillID)
	if current >= def.maxStacks {
		return false
	}
	player.SelectedSkills[skillID] = current + 1
	if skillID == "vital_growth" {
		player.MaxHealth += 15
		player.Health = min(player.MaxHealth, player.Health+15)
	}
	recalculateSkillDerivedStats(player)
	return true
}

func availableSkillDefinitions(player PlayerState) []skillDefinition {
	out := []skillDefinition{}
	for _, skill := range skillRegistry {
		if skillStacks(player, skill.id) < skill.maxStacks {
			out = append(out, skill)
		}
	}
	return out
}

func skillCardFromDefinition(player PlayerState, def skillDefinition) SkillCard {
	return SkillCard{
		ID:            def.id,
		Name:          def.name,
		Category:      def.category,
		Rarity:        def.rarity,
		Description:   def.description,
		Icon:          def.icon,
		MaxStacks:     def.maxStacks,
		CurrentStacks: skillStacks(player, def.id),
	}
}

func nextSkillSeed(seed uint32) uint32 {
	return seed*1664525 + 1013904223
}

func generateSkillCardOffer(player PlayerState, offerID string, seed uint32) (SkillCardOffer, bool) {
	ensurePlayerSkillState(&player)
	pool := availableSkillDefinitions(player)
	if len(pool) == 0 {
		return SkillCardOffer{}, false
	}
	chosen := []skillDefinition{}
	usedCategories := map[string]struct{}{}
	for len(chosen) < 3 && len(chosen) < len(pool) {
		remaining := []skillDefinition{}
		for _, skill := range pool {
			duplicate := false
			for _, selected := range chosen {
				if selected.id == skill.id {
					duplicate = true
					break
				}
			}
			if !duplicate {
				remaining = append(remaining, skill)
			}
		}
		source := []skillDefinition{}
		for _, skill := range remaining {
			if _, ok := usedCategories[skill.category]; !ok {
				source = append(source, skill)
			}
		}
		if len(source) == 0 {
			source = remaining
		}
		totalWeight := 0
		for _, skill := range source {
			totalWeight += skill.weight
		}
		seed = nextSkillSeed(seed)
		pick := int(seed % uint32(max(1, totalWeight)))
		selected := source[len(source)-1]
		for _, skill := range source {
			pick -= skill.weight
			if pick < 0 {
				selected = skill
				break
			}
		}
		chosen = append(chosen, selected)
		usedCategories[selected.category] = struct{}{}
	}
	cards := make([]SkillCard, 0, len(chosen))
	for _, def := range chosen {
		cards = append(cards, skillCardFromDefinition(player, def))
	}
	return SkillCardOffer{OfferID: offerID, PlayerID: player.ID, Cards: cards}, true
}

func skillStatsPayload(player PlayerState) map[string]any {
	return map[string]any{
		"hp": player.Health,
		"maxHp": player.MaxHealth,
		"xp": player.Relics,
		"level": player.Level,
		"atk": player.Damage,
		"damage": player.Damage,
		"attackSpeed": player.AttackSpeed,
		"hitRange": player.HitRange,
		"attackCooldownMs": player.AttackCooldownMs,
		"damageReduction": player.DamageReduction,
		"shield": player.Shield,
		"maxShield": player.MaxShield,
		"jumpPower": player.JumpPower,
		"airControl": player.AirControl,
		"extraJumps": player.ExtraJumps,
		"dashUnlocked": player.DashUnlocked,
		"pickupRadius": player.PickupRadius,
		"xpGainMultiplier": player.XPGainMultiplier,
		"killXpMultiplier": player.KillXPMultiplier,
		"selectedSkills": player.SelectedSkills,
	}
}
